ra-tls/libenclave-tls/enclave-tls.c

```diff
@@ -0,0 +1,295 @@
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	} else {
+#define DLSYM(fn)                                                                       \
+                do {                                                                    \
+                        ftab.fn = &fn;                                                  \
+                } while (0)
+
+		DLSYM(wolfSSL_free);
+		DLSYM(wolfSSL_CTX_set_verify);
+		DLSYM(create_key_and_x509);
+		DLSYM(wolfSSL_read);
+		DLSYM(wolfSSL_write);
+		DLSYM(wolfSSL_CTX_new);
+		DLSYM(printf_identity);
+		DLSYM(wolfTLSv1_2_client_method);
+		DLSYM(wolfSSL_X509_get_der);
+		DLSYM(wolfSSL_get_peer_certificate);
+		DLSYM(wolfSSL_connect);
+		DLSYM(wolfSSL_negotiate);
+		DLSYM(wolfSSL_set_fd);
+		DLSYM(wolfSSL_new);
+		DLSYM(wolfTLSv1_2_server_method);
+		DLSYM(wolfSSL_CTX_free);
+		DLSYM(wolfSSL_Cleanup);
+		DLSYM(wolfSSL_Debugging_ON);
+		DLSYM(wolfSSL_Debugging_OFF);
+		DLSYM(wolfSSL_Init);
+#undef DLSYM
+	}
+
+#ifdef DEBUG
+	ftab.wolfSSL_Debugging_ON();
+#else
+	ftab.wolfSSL_Debugging_OFF();
+#endif
+
+	ftab.wolfSSL_Init(ctx->quote_type);
+
+	WOLFSSL_CTX *ws_ctx;
+	if (ctx->is_server)
+		ws_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_server_method());
+	else
+		ws_ctx = ftab.wolfSSL_CTX_new(ftab.wolfTLSv1_2_client_method());
+	if (!ws_ctx) {
+		fprintf(stderr, "ERROR: failed to create WOLFSSL_CTX\n");
+		ftab.wolfSSL_Cleanup();
+		return NULL;
+	}
+
+	if (ctx->is_server) {
+		ftab.create_key_and_x509(ws_ctx);
+		if (ctx->mutual) {
+			ftab.wolfSSL_CTX_set_verify(ws_ctx,
+						    WOLFSSL_VERIFY_PEER |
+						    WOLFSSL_VERIFY_FAIL_IF_NO_PEER_CERT,
+						    cert_verify_callback);
+		}
+	} else {
+		if (ctx->mutual) {
+			ftab.create_key_and_x509(ws_ctx);
+		}
+		ftab.wolfSSL_CTX_set_verify(ws_ctx, SSL_VERIFY_PEER,
+					    cert_verify_callback);
+	}
+
+	ctx->ssl = ftab.wolfSSL_new(ws_ctx);
+	if (!ctx->ssl) {
+		fprintf(stderr, "ERROR: failed to create WOLFSSL object\n");
+		ftab.wolfSSL_CTX_free(ctx->ssl);
+		return NULL;
+	}
+
+	/* Attach wolfSSL to the socket */
+	ftab.wolfSSL_set_fd(ctx->ssl, ctx->sockfd);
+
+	return ctx;
+}
+
+int enclave_tls_negotiate(enclave_tls_ctx *ctx)
+{
+	if(ctx == NULL || ctx->ssl == NULL){
+		ftab.wolfSSL_free(ctx->ssl);
+		return -1;
+	}
+
+	int ret;
+
+	if (ctx->is_server) {
+		ret = ftab.wolfSSL_negotiate(ctx->ssl);
+	} else {
+		ret = ftab.wolfSSL_connect(ctx->ssl);
+	}
+	if (ret != SSL_SUCCESS) {
+		ftab.wolfSSL_free(ctx->ssl);
+		return ret;
+	}
+
+	if ((ctx->is_server && ctx->mutual) || !ctx->is_server) {
+		WOLFSSL_X509 *srvcrt =
+			ftab.wolfSSL_get_peer_certificate(ctx->ssl);
+
+		int derSz;
+		unsigned char *der = ftab.wolfSSL_X509_get_der(srvcrt, &derSz);
+		ftab.printf_identity(der, derSz);
+	}
+	return SSL_SUCCESS;
+}
+
+int enclave_tls_write(enclave_tls_ctx *ctx, const void *in, int sz)
+{
+	if(ctx == NULL || ctx->ssl == NULL){
```

Comment:
OK. I recognized this is an unwise way because we only need to know whether ctx is completely initialized or not. So plz add a bool field in ctx called initialized. Once enclave_tls_init() succeeds, set it to true. And all other enclave tls api just needs to check the validation of ctx and ctx->initialized.